daffodil-runtime1/src/main/scala/edu/illinois/ncsa/daffodil/processors/parsers/PackedDecimalParsers.scala

```diff
@@ -0,0 +1,108 @@
+ * LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+ * OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS WITH THE
+ * SOFTWARE.
+ */
+
+package edu.illinois.ncsa.daffodil.processors.parsers
+
+import edu.illinois.ncsa.daffodil.processors.Evaluatable
+import edu.illinois.ncsa.daffodil.schema.annotation.props.gen.LengthUnits
+import edu.illinois.ncsa.daffodil.schema.annotation.props.gen.YesNo
+import edu.illinois.ncsa.daffodil.processors.ElementRuntimeData
+import edu.illinois.ncsa.daffodil.util.DecimalUtils.packedToBigDecimal
+import java.lang.{ Long => JLong, Number => JNumber }
+import edu.illinois.ncsa.daffodil.processors.ParseOrUnparseState
+
+class PackedDecimalKnownLengthParser(e: ElementRuntimeData, signed: YesNo, binaryDecimalVirtualPoint: Int, val lengthInBits: Int)
+  extends PackedDecimalParserBase(e, signed, binaryDecimalVirtualPoint)
+  with HasKnownLengthInBits {
+}
+
+class PackedDecimalRuntimeLengthParser(val e: ElementRuntimeData, signed: YesNo, binaryDecimalVirtualPoint: Int, val lengthEv: Evaluatable[JLong], val lUnits: LengthUnits)
+  extends PackedDecimalParserBase(e, signed, binaryDecimalVirtualPoint)
+  with HasRuntimeExplicitLength {
+}
+
+abstract class PackedDecimalParserBase(e: ElementRuntimeData, signed: YesNo, binaryDecimalVirtualPoint: Int)
+  extends PrimParserObject(e) {
+
+  protected def getBitLength(s: ParseOrUnparseState): Int
+
+  def parse(start: PState): Unit = {
+    val nBits = getBitLength(start)
+    val dis = start.dataInputStream
+
+    if (!dis.isDefinedForLength(nBits)) {
+      PE(start, "Insufficient bits in data. Needed %d bit(s) but found only %d available.", nBits, dis.remainingBits.get)
+      return
+    }
+
+    val bigDec = packedToBigDecimal(dis.getByteArray(nBits), binaryDecimalVirtualPoint)
+    start.simpleElement.overwriteDataValue(bigDec)
+  }
+}
+
+class PackedIntegerRuntimeLengthParser(val e: ElementRuntimeData, signed: Boolean, val lengthEv: Evaluatable[JLong], val lUnits: LengthUnits)
+  extends PackedIntegerBaseParser(e, signed)
+  with HasRuntimeExplicitLength {
+}
+
+class PackedIntegerKnownLengthParser(e: ElementRuntimeData, signed: Boolean, val lengthInBits: Int)
+  extends PackedIntegerBaseParser(e, signed)
+  with HasKnownLengthInBits {
+}
+
+abstract class PackedIntegerBaseParser(e: ElementRuntimeData, signed: Boolean)
+  extends PrimParserObject(e) {
+
+  protected def getBitLength(s: ParseOrUnparseState): Int
+
+  def parse(start: PState): Unit = {
+    val nBits = getBitLength(start)
+    if (nBits == 0) return // zero length is used for outputValueCalc often.
+    val dis = start.dataInputStream
+
+    if (!dis.isDefinedForLength(nBits)) {
+      PE(start, "Insufficient bits in data. Needed %d bit(s) but found only %d available.", nBits, dis.remainingBits.get)
+      return
+    }
+
+    val int: JNumber =
+      if (signed) {
+        if (nBits > 64) { packedToBigDecimal(dis.getByteArray(nBits), 0) }
```

Comment:
Same comment as SL made above about (signed).

Repeat of this code suggests refactor into a base trait/class.